duck-type values responding to #to_hash or #to_ary

notEthan opened the pull request:
I work with JSON data wrapped in classes which behave like Hash and Array, which respond to #to_hash / #to_ary to support implicit conversion (as described in https://docs.ruby-lang.org/en/master/doc/implicit_conversion_rdoc.html )

this PR lets jmespath be used with these implicitly convertible objects. it is in draft because I have updated no tests or documentation, and may be incomplete in other ways. if an active maintainer can tell me it might be merged, I will bring it to completion.

notEthan committed:
duck-type values responding to #to_hash, #to_ary, #to_str rather than checking class

notEthan commented on lib/jmespath/nodes/function.rb:
```diff
@@ -145,8 +152,10 @@ def call(args)
-          if String === haystack || Array === haystack
-            haystack.include?(needle)
+          if haystack.respond_to?(:to_str)
+            haystack.to_str.include?(needle)
```
need to be more consistent in updating String/#to_str - I do here but not other places

mullermp commented:
Thanks for opening up a draft PR. I think what you're trying to do makes sense. What is the experience without any changes? - i.e. for your JSON data wrapped classes, are you able to call `to_a` or `to_h` on them and pass those raw converted values into jmespath?

alextwoods commented on lib/jmespath/nodes.rb:
```diff
@@ -5,10 +5,6 @@ class Node
       end
 
-      def hash_like?(value)
-        Hash === value || Struct === value
```
I believe this may be a breaking change as `Struct` does not implement `to_hash`.

notEthan replied:
yes, since Struct has no #to_hash, I changed the places which used this method to either call #to_hash or handle the Struct, handling both cases. by refactoring those there were no longer any calls to this method, so I removed it.

alextwoods replied:
Gotcha - agree it looks like Struct is being handled correctly where this was used.

alextwoods reviewed (commented):
Overall I think this is a reasonable change, but see comment re: `Struct`.

notEthan commented:
> What is the experience without any changes? - i.e. for your JSON data wrapped classes, are you able to call to_a or to_h on them and pass those raw converted values into jmespath?

I can pass the JSON that my classes wrap to jmespath, yes, but then lose the functionality that is added by the wrapper I have implemented. my goal is to search the wrapped document and get wrapped children out as the search result(s).

the wrapping I work with is associating JSON data with JSON schemas (per the specification at https://json-schema.org/) describing those data. so if I search within a document described by a schema, the resulting children get returned associated with subschemas that describe them. the library implementing this is here, in case that is of any interest: https://github.com/notEthan/jsi

if this is merged and released I'll add a method something like `#jmespath_search(expression)` to the wrapper class, to help with searching documents.

alextwoods commented on lib/jmespath/nodes/function.rb:
```diff
@@ -323,10 +333,12 @@ class ValuesFunction < Function
+          if value.respond_to?(:to_hash)
+            value.to_hash.values
+          elsif value.is_a?(Struct)
+            value.to_h.values
```
Can just call `values` on the struct instead of calling `to_h` first.

notEthan replied:
thanks

notEthan committed:
test handling of objects which are implicitly convertible to hash or array

notEthan commented:
I think I have the code in a good state for review, and have added a couple tests

mullermp commented:
Thanks. We'll review this shortly.

mullermp commented on spec/implicit_conversion_spec.rb:
```diff
@@ -0,0 +1,58 @@
+require 'spec_helper'
```
Do these tests cover all of the changes? Maybe something to consider is to run all existing tests but with wrappers too - if there's a least intrusive way of doing this.

alextwoods replied:
I've added CI back to this project and merged those changes in.    I've also ported the compliance tests into the implicit conversion test cases - there were a few failures in functions which I've fixed. 

There are a few other failures in the compliance tests w/ implicit conversion that I'm not positive how to address (they may be issues in my spec code as well) - could you take a look at those cases?

alextwoods committed:
Add compliance specs to the implicit conversion

alextwoods commented:
I've added CI back to this project and merged those changes in. I've also ported the compliance tests into the implicit conversion test cases - there were a few failures in functions which I've fixed.

There are a few other failures in the compliance tests w/ implicit conversion that I'm not positive how to address (they may be issues in my spec code as well) - could you take a look at those cases?

notEthan committed:
add Util.as_json to recursively handle comparison of complex objects

notEthan committed:
add test case for equality of complex items

notEthan committed:
add test case for array containing a complex item

notEthan commented on lib/jmespath/nodes/function.rb:
```diff
@@ -492,7 +509,7 @@ module CompareBy
       def compare_by(mode, *args)
         if args.count == 2
-          values = args[0]
+          values = args[0].to_ary
```
this one should convert #to_ary after the check `get_type(values) == ARRAY_TYPE`

notEthan commented:
thanks for the additions to the tests, that was definitely needed. I added and fixed a couple more to_arys, and added a `Util.as_json` for recursive comparison (this made redundant the unwrap method you'd added to the tests, so I removed that). added some test cases to the compliance json to cover code changes.
also rebased against main.

alextwoods reviewed (approved):
LGTM - great contribution and thanks for the extra testing!

notEthan commented:
glad to hear. thanks very much for your reviews and additions. I am quite happy to find active maintainers here, the project looked very quiet. also, aside, this is among the best-written libraries I've worked with or contributed to.
looking forward to integrating support when this is merged and gets released.

alextwoods commented:
@notEthan - Thanks again for a great contribution. I've released this as [version 1.5.0](https://rubygems.org/gems/jmespath/versions/1.5.0).

brianalexander commented:
For people searching google:

Any developers using the AWS SDK for Ruby v2 that has a dependency on jmespath, `Aws::EC2::Instance.block_device_mappings` returns an array of [InstanceBlockDeviceMapping](https://docs.aws.amazon.com/sdk-for-ruby/v2/api/Aws/EC2/Types/InstanceBlockDeviceMapping.html) that are now converted from Structs to Hashes implicitly.

To correct errors, update dot operator to bracket notation for value access.

Old way:  
```ruby
block_devices = Aws::EC2::Instance.new(instance_id, client: EC2_CLIENT_OBJECT).block_device_mappings
data_block_device = block_devices.find { |bd| bd.device_name == "/dev/xvdb" }
data_block_device.ebs.volume_id
```

New way:  
```ruby
block_devices = Aws::EC2::Instance.new(instance_id, client: EC2_CLIENT_OBJECT).block_device_mappings
data_block_device = block_devices.find { |bd| bd[:device_name] == "/dev/xvdb" }
data_block_device[:ebs][:volume_id]
```

mullermp commented:
That seems wrong. It should be a struct. We will investigate.

alextwoods commented:
Thanks for reporting that - I agree with @mullermp  - it should be a struct there.  I have confirmed this does NOT impact V3 of the SDK (note: [V2 of the SDK is no longer supported](https://aws.amazon.com/blogs/developer/deprecation-schedule-for-aws-sdk-for-ruby-v2/)).
